File: platform/sdk/nodejs/examples/t3-08-resume-partial.js
const { CloudClient } = require("../xuanji-sdk-cloud");

function main() {
  const client = new CloudClient();
  const bucket = "resume";
  const key = "resume/final.dat";

  // First attempt: upload 2 parts, abort (simulated failure)
  const r1 = client.createMultipartUpload(bucket, key);
  const uid1 = r1.upload_id;
  client.uploadPart(bucket, key, uid1, 1, Buffer.alloc(200, 1));
  client.uploadPart(bucket, key, uid1, 2, Buffer.alloc(200, 2));
  client.abortMultipartUpload(uid1);
  if (client.listMultipartUploads().count !== 0) throw new Error("aborted upload should be removed");

  // Second attempt: resume by creating new upload, upload all 4 parts
  const r2 = client.createMultipartUpload(bucket, key);
  const uid2 = r2.upload_id;
  const sizes = [200, 200, 300, 300];
  const parts = [];
  let total = 0;
  for (let i = 0; i < sizes.length; i++) {
    const n = i + 1;
    const data = Buffer.alloc(sizes[i], n & 0xFF);
    total += sizes[i];
    const r = client.uploadPart(bucket, key, uid2, n, data);
    if (!r.ok) throw new Error("part " + n + " failed");
    parts.push({ part_number: n, etag: r.etag });
  }
  const fin = client.completeMultipartUpload(bucket, key, uid2, parts);
  if (!fin.ok) throw new Error("complete failed");
  const obj = client.getObject(bucket, key);
  if (!obj.found) throw new Error("final obj missing");
  const size = (obj.data && obj.data.length) || 0;
  if (size !== total) throw new Error("size mismatch");
  if (client.listMultipartUploads().count !== 0) throw new Error("no uploads should remain after complete");
  console.log("XJ-OK: t3-08-resume-partial first_uid=" + uid1 + " second_uid=" + uid2 + " total_bytes=" + total + " etag=" + fin.etag);
}

if (require.main === module) main();
module.exports = main;

File: platform/sdk/python/examples/t3_08_resume_partial.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from xuanji_sdk_cloud import CloudClient

def main():
    client = CloudClient()
    bucket = "resume"
    key = "resume/final.dat"

    r1 = client.createMultipartUpload(bucket, key)
    uid1 = r1["upload_id"]
    client.uploadPart(bucket, key, uid1, 1, bytes([1] * 200))
    client.uploadPart(bucket, key, uid1, 2, bytes([2] * 200))
    client.abortMultipartUpload(uid1)
    if client.listMultipartUploads()["count"] != 0:
        raise RuntimeError("aborted upload should be removed")

    r2 = client.createMultipartUpload(bucket, key)
    uid2 = r2["upload_id"]
    sizes = [200, 200, 300, 300]
    parts = []
    total = 0
    for i, sz in enumerate(sizes):
        n = i + 1
        data = bytes([n & 0xFF] * sz)
        total += sz
        r = client.uploadPart(bucket, key, uid2, n, data)
        if not r["ok"]:
            raise RuntimeError(f"part {n} failed")
        parts.append({"part_number": n, "etag": r["etag"]})
    fin = client.completeMultipartUpload(bucket, key, uid2, parts)
    if not fin["ok"]:
        raise RuntimeError("complete failed")
    obj = client.getObject(bucket, key)
    if not obj["found"]:
        raise RuntimeError("final obj missing")
    data = obj["data"] if obj["data"] is not None else b""
    size = len(data)
    if size != total:
        raise RuntimeError("size mismatch")
    if client.listMultipartUploads()["count"] != 0:
        raise RuntimeError("no uploads should remain after complete")
    print(f"XJ-OK: t3_08_resume_partial first_uid={uid1} second_uid={uid2} total_bytes={total} etag={fin['etag']}")

if __name__ == "__main__":
    main()
